client/question/services/FeedbackGeneratorService.js

```diff
@@ -24,18 +24,26 @@ tie.factory('FeedbackGeneratorService', [
   'RUNTIME_ERROR_FEEDBACK_MESSAGES', 'WRONG_LANGUAGE_ERRORS', 'LANGUAGE_PYTHON',
   'CLASS_NAME_AUXILIARY_CODE', 'CLASS_NAME_SYSTEM_CODE', 'PARAGRAPH_TYPE_TEXT',
   'PARAGRAPH_TYPE_CODE', 'PARAGRAPH_TYPE_SYNTAX_ERROR',
-  'PYTHON_PRIMER_BUTTON_NAME', 'ERROR_COUNTER_LANGUAGE_UNFAMILIARITY',
-  'ERROR_COUNTER_SAME_RUNTIME', 'UNFAMILIARITY_THRESHOLD',
-  'FEEDBACK_CATEGORIES',
+  'PARAGRAPH_TYPE_OUTPUT', 'PYTHON_PRIMER_BUTTON_NAME',
+  'ERROR_COUNTER_LANGUAGE_UNFAMILIARITY', 'ERROR_COUNTER_SAME_RUNTIME',
+  'UNFAMILIARITY_THRESHOLD', 'FEEDBACK_CATEGORIES',
   function(
     FeedbackObjectFactory, TranscriptService, ReinforcementGeneratorService,
     CODE_EXECUTION_TIMEOUT_SECONDS, SUPPORTED_PYTHON_LIBS,
     RUNTIME_ERROR_FEEDBACK_MESSAGES, WRONG_LANGUAGE_ERRORS, LANGUAGE_PYTHON,
     CLASS_NAME_AUXILIARY_CODE, CLASS_NAME_SYSTEM_CODE, PARAGRAPH_TYPE_TEXT,
     PARAGRAPH_TYPE_CODE, PARAGRAPH_TYPE_SYNTAX_ERROR,
-    PYTHON_PRIMER_BUTTON_NAME, ERROR_COUNTER_LANGUAGE_UNFAMILIARITY,
-    ERROR_COUNTER_SAME_RUNTIME, UNFAMILIARITY_THRESHOLD,
-    FEEDBACK_CATEGORIES) {
+    PARAGRAPH_TYPE_OUTPUT, PYTHON_PRIMER_BUTTON_NAME,
+    ERROR_COUNTER_LANGUAGE_UNFAMILIARITY, ERROR_COUNTER_SAME_RUNTIME,
+    UNFAMILIARITY_THRESHOLD, FEEDBACK_CATEGORIES) {
+
+    /**
+     * Object used to keep track of how many corrective feedback (values)
+     * were generated for a given test suite (keys).
+     *
+     * @type {Object.<string, number>}
+     */
+    var attemptCounter = {};
```

Comment:
Two general design questions:

- If the user leaves the page and comes back, will this status be preserved? I'm guessing not, and that may be fine for now, but I just wanted to point that out.
- If the user gets an error for test 1 in suite A, and proceeds to the "expected output given" step, then fixes that error but their code is wrong for test 3 in suite A (despite passing test 1 in suite A), what should happen? This counter collapses all status data into a per-test-suite dict and it's not clear to me whether that correctly handles the case I've just described.

Reply:
- That's true. I agree it's probably fine for this PR though.
- Argh.. There was a regression check that was in there before; I must have accidentally deleted it when removing the hint related code. It checked whether the user failed on an earlier test case they had previously passed. I'll write it in again. When you say "test", are you referring to testCases as a whole or the individual input/allowedOutput pairs within a testCase?

Reply:
Will update per discussion.